_posts/2019-02-21-asp-net-core-routes-middleware.md

```diff
@@ -0,0 +1,186 @@
+
+### TL;DR
+
+Routing has traditionally been a construct confined within the ASP.NET ecosystem. As long as your code executes within that construct, you can read route data. However, with the proliferation of middleware that run before and after ASP.NET, the desire/need for an app.-wide, exposed routing system increased.
+
+## The 🗝️ to making this work
+
+### ASP.NET Core 2.2
+
+```csharp
+using Microsoft.AspNetCore.Internal;
+
+public class Startup
+{
+    public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+    {
+        app
+            .UseEndpointRouting() // Registering this prior to your middleware unlocks the ✨.
+            //.YourMiddleware()
+            .UseMvc();
+    }
+}
+```
+
+### ASP.NET Core 3.0 (preview-2)
+
+```csharp
+using Microsoft.AspNetCore.Builder;
+
+public class Startup
+{
+    public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+    {
+        app
+            .UseRouting() // Registering this prior to your middleware unlocks the ✨.
+            //.YourMiddleware()
+            .UseMvc();
+    }
+}
+```
+
+## Example
+
+*Imagine you're building a new blog engine...* 🤔
+
+💡 Maybe you want to pre-load a post via middleware so it's available anywhere down the chain. This way, none of other middleware is concerned with **how** to retrieve the post and can instead focus on whatever it does to augment the response.
+
+### Model
+
+```csharp
+public class Post
+{
+    public string Slug { get; set; }
+    public string Title { get; set; }
+}
+
+/// <summary>
+/// We need a derivative class to support scenarios where we can't find a `Post`-record based on {slug}.
+/// DI will throw an exception if we try to give it `null`.
+/// </summary>
+public class GenericPost : Post
+{ }
+```
+
+### Razor view
+
+```html
+@page "{slug}"
+@model PostModel
+
+<h1>@Model.Title</h1>
+...
+```
+
+### Razor view code
+```csharp
+public class PostModel : BaseModel
+{
+    public PostModel(Post post)
+    {
+        this.post = post;
+    }
+
+    public Post post { get; private set; }
+
+    public void OnGet()
+    { }
+}
+```
+
+### Middleware
+
+```csharp
+public class PostMiddleware
+{
+    public PostMiddlware(
+        RequestDelegate next,
+        IDatabase database)
+    )
+    {
+        this.next = next;
+        this.database = database;
+    }
+
+    private readonly IDatabase datbase;
+    private readonly RequestDelegate next;
+
+    public async Task InvokeAsync(HttpContext context)
+    {
+        var route = context.GetRouteData();
+
+        if (route.Values.TryGetValue("slug", out var routeSlug))
+        {
+            var slug = routeSlug.ToString();
+            var post = database.Fetch<Post>().FirstOrDefault(x => x.Slug == slug);
```

Comment:
For this example should we make this database call async?

Reply:
I couldn't tell if this was a :trollface: question, but made the change anyway. The database implementation is agnostic in the post–even though it's syntax mirrors `NPoco`.

Reply:
Nah I wasn't trolling. Just saw a database call that wasn't async inside an async method.